### ISSUE TITLE
Gradient Checkpointing for LongformerForSequenceClassification produces memory error

### ISSUE THREAD

**Issue by Weilin37** [allenai/longformer#80]
I am trying to include gradient checking for LongformerForSequenceClassification model. 

I first tested this on 512 tokens and it seems to produce a cuda out of memory error (GPU is K80). If I don't include gradient_checkpoint=True, then there's no error, with the rest of the code staying the same.

I have a multi-gpu setup but this breaks as well for single gpu.

`model = LongformerForSequenceClassification.from_pretrained('allenai/longformer-base-4096',
                                                                #gradient_checkpointing=True, # New to v3
                                                                num_labels=4)
    model = torch.nn.DataParallel(model) # Parallelize for multi GPU`

**Comment by ibeltagy** [comment]
K80 has 24GB of memory, right? I would expect this to work even without gradient checkpointing. What is the batch size you are using? also, can you share the full stacktrace of the error?

**Comment by Weilin37** [comment]
Yes. It works fine without gradient checkpoint on 512 tokens (and more of course). However, the only change I implemented is to set gradient_checkpointing = True and it breaks. I will post the stacktrace soon, but it points to the loss.backward() step.

**Comment by ibeltagy** [comment]
> I am using batch size = 16

You need to use batch size = 1. The 24GB of memory is enough for just 1 or 2 documents of length 4096. To increase the effective batch size, use gradient accumulation.

**Comment by Weilin37** [comment]
Thanks, but it does not explain why it would fail for 512 tokens using 4xK80's, or are you saying gradient checkpointing is only possible with a batch size of 1?

Here is my stacktrace:

```
---------------------------------------------------------------------------
RuntimeError Traceback (most recent call last)
<ipython-input-15-10796c70ef83> in <module>()
    11 ZPatient_labels_train = torch.tensor(ZPatient_labels_train)
     12 Train_dataset = TensorDataset(Input_ids_train, Attention_masks_train, ZPatient_labels_train)
---> 13 model_train = train_model(Train_dataset, epochs=1) 14 #Save model (optional) 15 #model_train.save_pretrained('./models/model_6_22')
 
<ipython-input-11-5b3244794b51> in train_model(train_dataset, epochs)
     87
     88 # Perform a backward pass to calculate the gradients.
---> 89 loss.backward()
     90
     91 # Clip the norm of the gradients to 1.0. This is to help prevent the "exploding gradients" problem.
~/anaconda3/envs/python3/lib/python3.6/site-packages/torch/tensor.py in backward(self, gradient, retain_graph, create_graph)
    196 products. Defaults to ``False``.
    197 """
--> 198 torch.autograd.backward(self, gradient, retain_graph, create_graph)
     199
     200 def register_hook(self, hook):
 
~/anaconda3/envs/python3/lib/python3.6/site-packages/torch/autograd/__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables)
     98 Variable._execution_engine.run_backward(
     99 tensors, grad_tensors, retain_graph, create_graph,
--> 100 allow_unreachable=True) # allow_unreachable flag
    101
    102
RuntimeError: CUDA out of memory. Tried to allocate 20.00 MiB (GPU 0; 11.17 GiB total capacity; 9.93 GiB already allocated; 7.56 MiB free; 10.83 GiB reserved in total by PyTorch)
```

**Comment by Weilin37** [comment]
Quick update: I set the batch size to 1 per GPU and it still gave me an OOM error on 512 tokens for multi-gpu

**Comment by ibeltagy** [comment]
can you try it on a single GPU just to test one thing at a time? It shouldn't OOM for a batch_size=1 and seqlen = 512 with or without gradient checkpointing

> but it does not explain why it would fail for 512 tokens using 4xK80's, or explain why removing gradient_checkpointing makes it work suddenly. 

yeah, this shouldn't happen. It should work with and without gradient checkpointing. This looks like a bug somewhere.  

Btw, it is recommended to use `DistributedDataParallel` and not use `DataParallel`. It could be an issue with `DataParallel`. IMO, the easiest way to setup multigpu with `DistributedDataParallel` is to use pytorch-lightning.

**Comment by ibeltagy** [comment]
> EDIT: Seems like my K80 only has 12gb per GPU

`batch_size=1, seqlen=4096, gradient_checkpointing=True` should work on 12gb gpu

**Comment by Weilin37** [comment]
Thanks. We tested it on a single V100 and it was able to do fine. The problem I guess is now the multi-gpu setup. We really can't get it to work after a few days of working on it, even with instructions in pytorch lightning. Is there a template code you can provide that can help us get started? We must be setting the DistributedDataParallel part wrong.

**Comment by ibeltagy** [comment]
What errors are you getting with DDP? you don't need more than `trainer = Trainer(gpus=8)`. The instructions [here](https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html) might be helpful as well.


> Is there a template code you can provide that can help us get started?

Well, our TriviaQA [script](https://github.com/allenai/longformer/blob/master/scripts/triviaqa.py) uses DDP so it might be helpful, but it is using an old version of PTL that I won't recommend, and it has lots of moving pieces that it doesn't work as a get-started example.

**Comment by Weilin37** [comment]
This is our error. I think we suspect that it could be an issue with jupyterlab. What do you think?

```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-21-0b32c3999a0c> in <module>()
      3 os.environ['MASTER_ADDR'] = 'localhost'
      4 os.environ['MASTER_PORT'] = '9000'
----> 5 mp.spawn(train_model, nprocs=8)
      6 #########################################################

~/anaconda3/envs/python3/lib/python3.6/site-packages/torch/multiprocessing/spawn.py in spawn(fn, args, nprocs, join, daemon, start_method)
    198                ' torch.multiprocessing.start_process(...)' % start_method)
    199         warnings.warn(msg)
--> 200     return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')

~/anaconda3/envs/python3/lib/python3.6/site-packages/torch/multiprocessing/spawn.py in start_processes(fn, args, nprocs, join, daemon, start_method)
    156 
    157     # Loop on join until it returns True or raises an exception.
--> 158     while not context.join():
    159         pass
    160 

~/anaconda3/envs/python3/lib/python3.6/site-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    111                 raise Exception(
    112                     "process %d terminated with exit code %d" %
--> 113                     (error_index, exitcode)
    114                 )
    115 

Exception: process 1 terminated with exit code 
```

**Comment by ibeltagy** [comment]
Looks like jupyterlab is messing up the spawning of the new processes. Please try regular python

**Comment by Weilin37** [comment]
Hi,

To follow up on this. I tried regular python and couldn't get it to work still. I think it is because the primary reason is that the distributed data parallel processing module isn't supported in Pytorch windows. Is it possible to implement this with just the regular data parallel processing module? At this point it seems like this would be the only way to make it work.

**Comment by Weilin37** [comment]
OK, I was able to run the script in jupyter notebook but it seems that LongformerForSequenceClassification is not compatible with Data Distributed Parallel.

I tried a simple example with 1 GPU:
```
dist.init_process_group(backend='nccl', init_method='env://', world_size=1, rank=0) #world_size is numGPUs*numNodes
    torch.manual_seed(seed_val)
    model = LongformerForSequenceClassification.from_pretrained('allenai/longformer-base-4096',
                                                                gradient_checkpointing=True,
                                                                num_labels=4)

    print(torch.cuda.get_device_properties(0).total_memory)
    torch.cuda.set_device(gpu)
    model.cuda(gpu)
    
    #device = torch.device("cuda:0")
    #model.to(device) # Move to GPU
    
    batch_size = 1 # CHANGE BATCH SIZE HERE
    epochs = 1 # CHANGE NUM EPOCHS HERE
    
    optimizer = AdamW(model.parameters(),
                  lr = 2e-5,
                  eps = 1e-8
                )

    model = nn.parallel.DistributedDataParallel(model, find_unused_parameters=False)
    

    train_sampler = torch.utils.data.distributed.DistributedSampler(train_dataset,
                                                                    num_replicas=1, # World size
                                                                    rank=0) # Only one node, so rank=gpu
    
    train_dataloader = torch.utils.data.DataLoader(dataset=train_dataset,
                                               batch_size=batch_size,
                                               shuffle=False,
                                               num_workers=0,
                                               pin_memory=True,
                                               sampler=train_sampler)
```

and got this error.

```
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by (1) passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`; (2) making sure all `forward` function outputs participate in calculating loss. If you already have done the above two steps, then the distributed data-parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
```

Searching the internet, I ran this code after the first backwards:

```
b_input_ids = batch[0].cuda(gpu)
            b_input_mask = batch[1].cuda(gpu)
            b_labels = batch[2].cuda(gpu)

            model.zero_grad()        

            loss, logits = model(b_input_ids, token_type_ids=None, attention_mask=b_input_mask, labels=b_labels)

            loss = loss.mean()
            total_train_loss += loss.item()

            loss.backward()
            
            # check parameters with no grad
            for n, p in model.named_parameters():
                if p.grad is None and p.requires_grad is True:
                    print('No forward parameters:', n, p.shape)
```

And it printed layers in the model that was not part of the forward step:

```
No forward parameters: module.longformer.pooler.dense.weight torch.Size([768, 768])
No forward parameters: module.longformer.pooler.dense.bias torch.Size([768])
```

So I think this is where the road ends. There are two layers within LongformerForSequenceClassification that prevents training in a multi-gpu setting. I get this error even after turning off gradient checkpointing.

Any advice on how to move forward would be much appreciated!

**Comment by matt-peters** [comment]
Try replacing `model = nn.parallel.DistributedDataParallel(model, find_unused_parameters=False)` with `model = nn.parallel.DistributedDataParallel(model, find_unused_parameters=True)` in your code, that usually fixes this particular error.

**Comment by Weilin37** [comment]
> Try replacing `model = nn.parallel.DistributedDataParallel(model, find_unused_parameters=False)` with `model = nn.parallel.DistributedDataParallel(model, find_unused_parameters=True)` in your code, that usually fixes this particular error.

I think there is a thread specifying that using find_unsused_paramters=True is incompatible with gradient checkpointing, so that's a no go. 

https://github.com/huggingface/transformers/pull/4659

**Comment by ibeltagy** [comment]
@Weilin37 , use `find_unused_parameters=False` and remove `module.longformer.pooler.dense.weight` and `module.longformer.pooler.dense.bias`  from your model.
